### PR TITLE
KAFKA 2578 Client Metadata internal state should be synchronized

### PR DESCRIPTION
**Pull request by eribeiro**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 1]
```diff
@@ -43,7 +43,7 @@
     private Cluster cluster;
     private boolean needUpdate;
     private final Set<String> topics;
-    private final List<Listener> listeners;
+    private final Set<Listener> listeners;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Why did you change this to a `Set`? It seems unrelated to the ticket.

Comment (eribeiro) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: @ijuma Yeah, my fault. Seemed logical to be a `Set` instead of a `List` so that we don't have the same listener duplicated. But I need to stop doing this kind of unrelated change. Excuse me. :) Gonna revert now and make sure I stop doing this forever.

**Comment by hachikuji** [comment]
LGTM. We probably should have just made this a MINOR patch to keep jira noise down, but oh well.

**Comment by eribeiro** [comment]
Well, the failed test (kafka.network.SocketServerTest > testSocketsCloseOnShutdown) looks unrelated to the issue.

**Comment by hachikuji** [comment]
@eribeiro Looks like there's already a jira for that test case (KAFKA-2398).

**Comment by ewencp** [comment]
@eribeiro Doesn't build for me since `Set` is still used in this file. Otherwise LGTM.

**Comment by ewencp** [comment]
@eribeiro We've resolved this (with attribution maintained) in https://github.com/apache/kafka/pull/659. When you have a chance, could you close out this PR?

**Comment by eribeiro** [comment]
@ewencp Hi Ewen, sorry for the delay*, closing now. Thanks for cherry-picking!
- I have been absent of Kafka project for some time now, coming back again :)
